jetty-core/jetty-siwe/src/main/java/org/eclipse/jetty/security/siwe/EthereumAuthenticator.java

```diff
@@ -0,0 +1,771 @@
+                    signature = parts.getFirst("signature").getContentAsString(StandardCharsets.ISO_8859_1);
+                    message = parts.getFirst("message").getContentAsString(StandardCharsets.ISO_8859_1);
+                     */
+                    throw new ServerAuthException("Unsupported mime type: " + mimeType);
+                }
+                default -> throw new ServerAuthException("Unsupported mime type: " + mimeType);
+            };
+
+            // The browser may convert LF to CRLF, EIP4361 specifies to only use LF.
+            message = message.replace("\r\n", "\n");
+            return new SignedMessage(message, signature);
+        }
+        catch (Throwable t)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("error reading SIWE message and signature", t);
+            sendError(request, response, callback, t.getMessage());
+            return null;
+        }
+    }
+
+    protected AuthenticationState handleNonceRequest(Request request, Response response, Callback callback)
+    {
+        String nonce = createNonce(request.getSession(false));
+        ByteBuffer content = BufferUtil.toBuffer("{ \"nonce\": \"" + nonce + "\" }");
+        response.write(true, content, callback);
+        return AuthenticationState.CHALLENGE;
+    }
+
+    private boolean validateSignInWithEthereumToken(SignInWithEthereumToken siwe, SignedMessage signedMessage, Request request, Response response, Callback callback)
+    {
+        Session session = request.getSession(false);
+        if (siwe == null)
+        {
+            sendError(request, response, callback, "failed to parse SIWE message");
+            return false;
+        }
+
+        try
+        {
+            siwe.validate(signedMessage, nonce -> redeemNonce(session, nonce), _schemes, _domains, _chainIds);
+        }
+        catch (Throwable t)
+        {
+            sendError(request, response, callback, t.getMessage());
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public AuthenticationState validateRequest(Request request, Response response, Callback callback) throws ServerAuthException
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("validateRequest({},{})", request, response);
+
+        String uri = request.getHttpURI().toString();
+        if (uri == null)
+            uri = "/";
+
+        try
+        {
+            Session session = request.getSession(false);
+            if (session == null)
+            {
+                session = request.getSession(true);
+                if (session == null)
+                {
+                    sendError(request, response, callback, "session could not be created");
+                    return AuthenticationState.SEND_FAILURE;
+                }
+            }
+
+//            boolean sessionIdFromCookie = (Boolean)request.getAttribute("org.eclipse.jetty.session.sessionIdFromCookie");
```

Comment:
I would still like to keep this check but there is no API to actually check this.
I had a go at doing it but reverted it because it wasn't working very well.
Probably something for a different PR anyway.

I removed this code and replaced with a TODO.